src/DelayBucket.cc

```diff
@@ -25,29 +27,42 @@ DelayBucket::stats(StoreEntry *entry)const
 void
 DelayBucket::update(DelaySpec const &rate, int incr)
 {
-    if (rate.restore_bps != -1 &&
-            (level() += rate.restore_bps * incr) > rate.max_bytes)
-        level() = rate.max_bytes;
+    // make sure an undefined NaturalProduct() result (below) implies overflow
+    if (rate.restore_bps < 0)
+        return;
+
+    const auto oldLevel = level();
+    if (const auto delta = IncreaseProduct(rate.restore_bps, incr)) {
+        if (const auto newLevel = IncreaseSum(oldLevel, delta.value())) {
+            level() = newLevel.value();
+            return;
+        }
+    }
+    // TODO: level() and rate.max_bytes should have the same type
+    const auto maxLevel = IncreaseSum(BucketLevel(0), rate.max_bytes);
+    assert(maxLevel);
+    level() = maxLevel.value();
 }
 
-int
-DelayBucket::bytesWanted(int minimum, int maximum) const
+DelayBucket::BucketLevel
+DelayBucket::bytesWanted(BucketLevel minimum, BucketLevel maximum) const
 {
-    int result = max(minimum, min(maximum, level()));
+    BucketLevel result = max(minimum, min(maximum, level()));
     return result;
 }
 
 void
-DelayBucket::bytesIn(int qty)
+DelayBucket::bytesIn(BucketLevel qty)
 {
     level() -= qty;
 }
 
 void
 DelayBucket::init(DelaySpec const &rate)
 {
-    level() = (int) (((double)rate.max_bytes *
-                      Config.Delay.initial) / 100);
+    const auto initialLevel = IncreaseSum(BucketLevel(0), rate.max_bytes * (Config.Delay.initial/100.));
```

Comment:
Hm... Are we missing an is_integral static assert in summation functions? Or do they actually correctly support floating point arguments?!

Reply:
I think they support floating point because there are actually two versions of IncreaseSum() depending on AllUnsigned<S,T> type.

Reply:
> is_specialized would describe this intention a little better

I do not see why you think that way. We want max() to be meaningful. is_bounded implies max() is meaningful. is_specialized does not imply max() is meaningful. There are realistic examples of (correct) types for which is_specialized is true, is_bounded is false, and max() might be present but meaningless.

> The compiler would catch us if the specialization lacks max() - because default max() is not generated in this case

Yes, but the problem we are trying to solve is not the lack of max() -- we do not need to add any static asserts to enforce that because the compiler will fail to compile our max()-using code if max() is outright missing! The problem we are actually trying to solve here is a type that has a meaningless max(). Given how STL _default_ is a meaningless max(), I would not be surprised if folks add meaningless max() definitions to other type specializations for some valid reason (unknown to me and speculated about by others[^1]).

[^1]: Like Bo Persson's "without `if constexpr`, it is often an advantage if both paths will compile" at 
https://stackoverflow.com/questions/35575276/why-does-stdnumeric-limitssecondsmax-return-0#comment59734406_35575526

> for a user-defined type, Sum methods would force that type to add a numeric_limits specialization with is_bounded=true

I do not like the "force" wording/implications here (Sum should not and may not be able to force others), but, ignoring that, this is an argument for is_bounded (more than it is an argument for is_specialized).

Reply:
> In summary, I suggest adding is_exact and is_bounded

I added these new requirements (and is_enum) at d6f3cf6. If we are happy with them, we can apply them to the official PR869.